### DISCUSSION TITLE
"stroke-dashoffset" works fine the the lovelact card preview, but no once you hit "save"

### DISCUSSION THREAD

**Discussion by Aephir** [ExperienceLovelace/ha-floorplan#257] [Q&A]
Specifically, I have this circle:

```
    <circle
       style="display:inline;fill-opacity:1;fill-rule:nonzero;stroke:#323232;stroke-width:6;stroke-linejoin:round;stroke-miterlimit:4;stroke-dasharray:none;stroke-opacity:1"
       id="real_circle"
       cx="271.52206"
       cy="85.024849"
       r="28.786091" />
```

I need the circle to hide X % (`100% - entity_state`, which is charge of a phone) of what's behind. To do that, I put in my floorplan.yaml:

```
                  style: |
                    >
                    return `
                      stroke-dasharray: 162 162;
                      stroke-dashoffset: ${-(162*(entity.state/100))};
                      transform-box: fill-box;
                      transform-origin: center;
                      transform: rotate(180deg);
                    `;
```
(162 being because the "real" radius is `r_real = r - (stroke-width/2) = 28.786091 - (6/2) = 25.786091`. Meaning the circumference is `c = 2 * pi * r_real ~ 162`).

In the preview (editing the lovelace card) it looks as expected (I had the charge at 66%):
<img width="178" alt="Screenshot 2022-10-23 at 21 38 24" src="https://user-images.githubusercontent.com/20152788/197414454-8a9dd830-5364-4d2d-a7bb-f477bd3e3c67.png">

But once I click save, it looks all wonky:
<img width="237" alt="Screenshot 2022-10-23 at 21 38 37" src="https://user-images.githubusercontent.com/20152788/197414473-8663bb32-53eb-46dc-be38-ab31b9d09597.png">

and I have no clue why? It looks like it has done some scaling, since both the stroke-dashoffset and circle stroke-width (cmpared too what's behind) is off. But if I use the Firefox developer tools, the size, `stroke-dasharray`, and `stroke-dashoffset` of the circle are all what I set them to, so I don't see where the scaling has happened.

So basically, I'm completely inn the dark about what this behavior is, and how to fix this.

Any suggestions would be greatly appreciated.

**Comment by OzGav** [comment]
You are forging a new path here so all I can suggest is try using chrome and see if that reveals anything different. I do know Firefox has some caching issues that sometimes cause weird results.

**Reply by Aephir** [reply]
Unfortunately, it's exactly the same with Firefox, Chrome, Vivaldi, Safari, Epic, and a few others. The only one that's slightly different (though still not correct) is the Tor browser.

Here, it looks like this instead:
<img width="156" alt="Screenshot 2022-10-24 at 09 16 42" src="https://user-images.githubusercontent.com/20152788/197469176-3a4d9814-4d71-4df7-8f82-9886f4c0d5c6.png">

And all the browser's dev tools show that size and stroke-dashoffset are the same as I have set.

**Reply by OzGav** [reply]
I will try and set it myself and see what happens. Have you made it a different colour so you can see exactly what is happening and then try and fudge it so it works? Clunky I know…..

**Reply by Aephir** [reply]
Yeah, I've tried quite a few "random" numbers, but can't seem to see anything sensible. Let me know if you want me to send any files or additional code!

**Comment by Aephir** [comment]
So, I think it might relate to this part of the css file:

```
svg, svg * {
  vector-effect: non-scaling-stroke !important;
  pointer-events: all !important;
}
```

If I delete it, is seems to work. I have the `vector-effect:` part from the examples, ad I'm not sure whether it will end up messing anything up eventually if I remove it?

I tried just overwriting for the specific object from the floorplan.yaml 

```
                    >
                    return `
                      stroke-dasharray: 162 162;
                      stroke-dashoffset: ${-(162*(.66))};
                      transform-box: fill-box;
                      transform-origin: center;
                      transform: rotate(180deg);
                      vector-effect: none;
                    `;
```

but that didn't work. Any idea how to apply that on a single object, while retaining the `  vector-effect: non-scaling-stroke !important;` in the css file?

**Reply by OzGav** [reply]
To be honest I had no idea what vector-effect did until 2 mins ago. Looking here though https://codepen.io/rlchy/pen/MaoJXQ

You can play with the HTML and delete that property and observe what happens to the circles. Only the largest circle is really noticeable. It just stops scaling the stroke width if the object is transformed in some way. So I can't see any reason not to remove it. Maybe your underlying image is the thing that is actually getting distorted by the browser? In any case I suggest removing it and then viewing it on a variety of devices and zoom in and out and see if you get the behaviour you are after

A bit more info here https://www.stefanjudis.com/today-i-learned/how-to-disable-scaling-for-stroke-width/

P.S. alternatively you could keep the entry in the svg section and try adding `vector-effect="none" !important` to your circles, that is, just try adding the `!important` bit

**Reply by exetico** [reply]
> Any idea how to apply that on a single object, while retaining the vector-effect: non-scaling-stroke !important; in the css file?

I've not checked all the information in your question, but `!important` on the dom-element itself (`<circle style"whatever!important">`), will be "more important" than `!important` in the CSS-file.

In your CSS file, you can also create a rule like:
```css
svg circle#real_circle{
    vector-effect: none!important;
}
```

Which will be higher prioritized than `svg *` if I remember correct. I normally prevents the usage of `*`. Using `svg *` is a real no-no if other options is possible, if you're trying to write clean and "understandable" code.  I'll suggest you to define the proper selectors, whenever possible.

I've actually talked about the usage of `!important` here:
https://www.youtube.com/watch?v=oqO4Vx6U1Mo

Here's a link about prioritization:
https://www.w3schools.com/css/tryit.asp?filename=trycss_specificity2

When I'm styling with CSS on my primary work, I normally prevent the usage of `!important` whenever possible 😄 ...

**Reply by Aephir** [reply]
Thanks @OzGav & @exetico . I have to admit, I really know (almost) nothing about css, I just copied the examples and started working from them. So I probably won't remove all the dozens of `!important` and `*` until I have time to properly understand the impacts (and what a better solution is for each case).

It doesn't seem to do what I expect if I keep 

```
svg, svg * {
  vector-effect: non-scaling-stroke !important;
  pointer-events: all !important;
}

svg circle#battery_circle.p30_pro_battery_level{
    vector-effect: none!important;
}
```

(`battery_circle.p30_pro_battery_level` being the name of the circle in my real floorplan.svg), so for now I'll probably just delete it.

With the `vector-effect: none!important;` for the specific circle):
<img width="119" alt="Screenshot 2022-10-31 at 09 38 40" src="https://user-images.githubusercontent.com/20152788/198966896-41ed143d-115c-4972-8557-18c0fd8908c8.png">


And when just deleting everything about `vector-effect:`:
<img width="115" alt="Screenshot 2022-10-31 at 09 40 47" src="https://user-images.githubusercontent.com/20152788/198966952-7685736d-e631-4cb0-80f5-c0e1ea4c4261.png">

But thanks, at least there's one way to make this work now.

**Reply by exetico** [reply]
Np!

Next time, consider to create a "test"-pack for us 😃.

Btw, I like the graphics!

**Reply by Aephir** [reply]
Thanks, I'll share the svg and code [here](https://community.home-assistant.io/t/smarter-way-to-conditionally-make-100-elements-clear-opaque-based-on-percentage-of-sensor-state/473680/6), I just need to decide on where to host the svg since it becomes ~43000 lines, and the forum only allows 32000.

**Reply by Aephir** [reply]
*characters. Not lines:)

**Reply by exetico** [reply]
Thank you for sharing your solution :smile:
